icu4c/source/common/localebuilder.cpp

```diff
@@ -0,0 +1,395 @@
+                    }
+                }
+                to->setKeywordValue(key, value.data(), error);
+            }
+        }
+    }
+}
+
+void static
+_clearUAttributesAndKeyType(Locale* locale, UErrorCode& error)
+{
+    // Clear Unicode attributes
+    locale->setKeywordValue(kAttributeKey, "", error);
+
+    // Clear all Unicode keyword values
+    StringEnumeration* str_enum = locale->createUnicodeKeywords(error);
+    if (U_SUCCESS(error) && str_enum != nullptr) {
+        std::unique_ptr<icu::StringEnumeration> iter(str_enum);
+        const char* key;
+        while (U_SUCCESS(error) &&
+               nullptr != (key = iter->next(nullptr, error))) {
+            locale->setUnicodeKeywordValue(key, nullptr, error);
+        }
+    }
+}
+
+static void
+_setUnicodeExtensions(Locale* locale, const CharString& value, UErrorCode& error)
+{
+    // Add the unicode extensions to extensions_
+    CharString locale_str("und-u-", error);
+    locale_str.append(value, error);
+    _copyExtensions(
+        Locale::forLanguageTag(locale_str.data(), error),
+        locale, false, error);
+}
+
+LocaleBuilder& LocaleBuilder::setExtension(char key, StringPiece value)
+{
+    if (U_FAILURE(status_)) return *this;
+    if (!UPRV_ISALPHANUM(key)) {
+        status_ = U_ILLEGAL_ARGUMENT_ERROR;
+        return *this;
+    }
+    CharString value_str(value, status_);
+    char* p = value_str.data();
+    for (int32_t i = 0; i < value_str.length(); i++, p++) {
+        *p = (*p == '_') ? '-' : uprv_tolower(*p);
+    }
+    if (value_str.length() > 0 &&
+        !_isExtensionSubtags(key, value_str.data(), value_str.length())) {
+        status_ = U_ILLEGAL_ARGUMENT_ERROR;
+        return *this;
+    }
+    if (extensions_ == nullptr) {
+        extensions_ = new Locale();
+    }
+    if (uprv_tolower(key) != 'u') {
+        // for t, x and others extension.
+        extensions_->setKeywordValue(StringPiece(&key, 1), value_str.data(),
+                                     status_);
+        return *this;
+    }
+    _clearUAttributesAndKeyType(extensions_, status_);
+    if (value.length() > 0) _setUnicodeExtensions(extensions_, value_str, status_);
+    return *this;
+}
+
+LocaleBuilder& LocaleBuilder::setUnicodeLocaleKeyword(
+      StringPiece key, StringPiece type)
+{
+    if (U_FAILURE(status_)) return *this;
+    if (!ultag_isUnicodeLocaleKey(key.data(), key.length()) ||
+        (type.length() > 0 &&
+         !ultag_isUnicodeLocaleType(type.data(), type.length()))) {
+      status_ = U_ILLEGAL_ARGUMENT_ERROR;
+      return *this;
+    }
+    if (extensions_ == nullptr) extensions_ = new Locale();
+    extensions_->setUnicodeKeywordValue(key, type, status_);
+    return *this;
+}
+
+LocaleBuilder& LocaleBuilder::addUnicodeLocaleAttribute(
+    StringPiece value)
+{
+    CharString value_str(value, status_);
+    if (U_FAILURE(status_)) return *this;
+    char* p = value_str.data();
+    for (int32_t i = 0; i < value_str.length(); i++, p++) {
```

Comment:
Please make a helper function that takes a non-const char* and length and does the _- and lowercasing -- and replace each of these loops.